File: app/src/main/java/com/mirash/passkeeper/activity/CredentialsEditActivity.java
package com.mirash.passkeeper.activity;

import android.os.Bundle;
import android.text.Editable;
import android.text.TextWatcher;
import android.view.Menu;
import android.view.MenuItem;

import androidx.annotation.Nullable;
import androidx.appcompat.app.ActionBar;
import androidx.appcompat.app.AppCompatActivity;
import androidx.lifecycle.Observer;
import androidx.lifecycle.ViewModelProvider;

import com.mirash.passkeeper.Const;
import com.mirash.passkeeper.R;
import com.mirash.passkeeper.db.Credentials;
import com.mirash.passkeeper.model.CredentialsModel;
import com.mirash.passkeeper.view.StyledTextInputLayout;

/**
 * @author Mirash
 */
public class CredentialsEditActivity extends AppCompatActivity implements Observer<Credentials> {
    private StyledTextInputLayout titleInputLayout;
    private StyledTextInputLayout linkInputLayout;
    private StyledTextInputLayout loginInputLayout;
    private StyledTextInputLayout passwordInputLayout;
    private StyledTextInputLayout pinInputLayout;
    private StyledTextInputLayout detailsInputLayout;
    private CredentialsEditViewModel model;
    private MenuItem saveMenuItem;

    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_credentials_edit);
        ActionBar actionBar = getSupportActionBar();
        if (actionBar != null) {
            actionBar.setDisplayHomeAsUpEnabled(true);
        }
        titleInputLayout = findViewById(R.id.item_title_input);
        linkInputLayout = findViewById(R.id.item_link_input);
        loginInputLayout = findViewById(R.id.item_login_input);
        passwordInputLayout = findViewById(R.id.item_password_input);
        pinInputLayout = findViewById(R.id.item_pin_input);
        detailsInputLayout = findViewById(R.id.item_details_input);

        model = new ViewModelProvider(this).get(CredentialsEditViewModel.class);
        Bundle args = getIntent().getExtras();
        if (args != null && args.containsKey(Const.KEY_ID)) {
            int id = getIntent().getIntExtra(Const.KEY_ID, 0);
            model.setCredentialsId(id);
            model.getCredentialsLiveData().observe(this, this);
        }
        initSaveButtonStateObserver();
    }

    @Override
    protected void onDestroy() {
        super.onDestroy();
        model.removeCredentialsObserver(this);
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        getMenuInflater().inflate(model.getCredentialsId() == null ? R.menu.credentials_save_menu : R.menu.credentials_edit_menu, menu);
        saveMenuItem = menu.findItem(R.id.credentials_edit_save);
        model.getSaveButtonEnableStateLiveData().observe(this, enabled -> {
            saveMenuItem.setEnabled(enabled);
            saveMenuItem.getIcon().setAlpha(enabled ? 255 : 78);
        });
        return super.onCreateOptionsMenu(menu);
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        int id = item.getItemId();
        if (id == android.R.id.home) {
            onBackPressed();
            return true;
        } else if (id == R.id.credentials_edit_save) {
            saveCredentials();
        } else if (id == R.id.credentials_edit_delete) {
            model.deleteCredentials();
            onBackPressed();
        }
        return super.onOptionsItemSelected(item);
    }

    @Override
    public void onChanged(Credentials credentials) {
        if (credentials == null) return;
        titleInputLayout.setTextWithNoAnimation(credentials.getTitle());
        linkInputLayout.setTextWithNoAnimation(credentials.getLink());
        loginInputLayout.setTextWithNoAnimation(credentials.getLogin());
        passwordInputLayout.setTextWithNoAnimation(credentials.getPassword());
        pinInputLayout.setTextWithNoAnimation(credentials.getPin());
        detailsInputLayout.setTextWithNoAnimation(credentials.getDetails());
    }

    private void saveCredentials() {
        CredentialsModel credentialsModel = new CredentialsModel();
        credentialsModel.setTitle(titleInputLayout.getText());
        credentialsModel.setLink(linkInputLayout.getText());
        credentialsModel.setLogin(loginInputLayout.getText());
        credentialsModel.setPassword(passwordInputLayout.getText());
        credentialsModel.setPin(pinInputLayout.getText());
        credentialsModel.setDetails(detailsInputLayout.getText());
        model.saveCredentials(credentialsModel);
        onBackPressed();
    }

    private void initSaveButtonStateObserver() {
        titleInputLayout.addTextChangedListener(new TextWatcher() {
            @Override
            public void beforeTextChanged(CharSequence charSequence, int i, int i1, int i2) {
            }

            @Override
            public void onTextChanged(CharSequence charSequence, int i, int i1, int i2) {
            }

            @Override
            public void afterTextChanged(Editable editable) {
                model.setFillState(CredentialsEditViewModel.INDEX_TITLE, editable.length() > 0);
            }
        });
        linkInputLayout.addTextChangedListener(new TextWatcher() {
            @Override
            public void beforeTextChanged(CharSequence charSequence, int i, int i1, int i2) {
            }

            @Override
            public void onTextChanged(CharSequence charSequence, int i, int i1, int i2) {
            }

            @Override
            public void afterTextChanged(Editable editable) {
                model.setFillState(CredentialsEditViewModel.INDEX_LINK, editable.length() > 0);
            }
        });
        loginInputLayout.addTextChangedListener(new TextWatcher() {
            @Override
            public void beforeTextChanged(CharSequence charSequence, int i, int i1, int i2) {
            }

            @Override
            public void onTextChanged(CharSequence charSequence, int i, int i1, int i2) {
            }

            @Override
            public void afterTextChanged(Editable editable) {
                model.setFillState(CredentialsEditViewModel.INDEX_LOGIN, editable.length() > 0);
            }
        });
        passwordInputLayout.addTextChangedListener(new TextWatcher() {
            @Override
            public void beforeTextChanged(CharSequence charSequence, int i, int i1, int i2) {
            }

            @Override
            public void onTextChanged(CharSequence charSequence, int i, int i1, int i2) {
            }

            @Override
            public void afterTextChanged(Editable editable) {
                model.setFillState(CredentialsEditViewModel.INDEX_PASSWORD, editable.length() > 0);
            }
        });
        pinInputLayout.addTextChangedListener(new TextWatcher() {
            @Override
            public void beforeTextChanged(CharSequence charSequence, int i, int i1, int i2) {
            }

            @Override
            public void onTextChanged(CharSequence charSequence, int i, int i1, int i2) {
            }

            @Override
            public void afterTextChanged(Editable editable) {
                model.setFillState(CredentialsEditViewModel.INDEX_PIN, editable.length() > 0);
            }
        });
    }
}

File: app/src/main/java/com/mirash/familiar/activity/edit/credentials/CredentialsEditActivity.kt
package com.mirash.familiar.activity.edit.credentials

import android.os.Bundle
import androidx.appcompat.widget.Toolbar
import androidx.lifecycle.Observer
import androidx.lifecycle.ViewModelProvider
import com.mirash.familiar.activity.edit.base.BaseEditActivity
import com.mirash.familiar.databinding.ActivityCredentialsEditBinding
import com.mirash.familiar.db.Credentials
import com.mirash.familiar.db.RepositoryProvider
import com.mirash.familiar.db.User
import com.mirash.familiar.model.credentials.CredentialsModel
import com.mirash.familiar.model.credentials.ICredentials
import com.mirash.familiar.tool.KEY_POSITION
import com.mirash.familiar.user.UserControl

/**
 * @author Mirash
 */
class CredentialsEditActivity :
    BaseEditActivity<ICredentials, Credentials, CredentialsEditViewModel, ActivityCredentialsEditBinding>() {

    override fun applyIntentExtras(bundle: Bundle) {
        super.applyIntentExtras(bundle)
        val position = intent.getIntExtra(KEY_POSITION, 0)
        model.credentialsPosition = position
    }

    override fun onCreateNewDataMode() {
        val liveData = RepositoryProvider.userRepository.getById(UserControl.userId)
        liveData.observe(this, object : Observer<User> {
            override fun onChanged(value: User) {
                liveData.removeObserver(this)
                binding.itemEmailInput.let {
                    if (it.text.isNullOrEmpty()) it.text = value.email
                }
                binding.itemPhoneInput.let {
                    if (it.text.isNullOrEmpty()) it.text = value.phone
                }
            }
        })
    }

    override fun initViewBinding(): ActivityCredentialsEditBinding =
        ActivityCredentialsEditBinding.inflate(layoutInflater)

    override fun initViewModel(): CredentialsEditViewModel =
        ViewModelProvider(this)[CredentialsEditViewModel::class.java]

    override fun toolbar(): Toolbar = binding.credentialsToolbar

    override fun createData(): ICredentials {
        val credentialsModel = CredentialsModel()
        credentialsModel.title = binding.itemTitleInput.text
        credentialsModel.link = binding.itemLinkInput.text
        credentialsModel.login = binding.itemLoginInput.text
        credentialsModel.email = binding.itemEmailInput.text
        credentialsModel.phone = binding.itemPhoneInput.text
        credentialsModel.password = binding.itemPasswordInput.text
        credentialsModel.pin = binding.itemPinInput.text
        credentialsModel.details = binding.itemDetailsInput.text
        credentialsModel.position = model.credentialsPosition
        return credentialsModel
    }

    override fun onChanged(value: Credentials) {
        binding.itemTitleInput.setTextWithNoAnimation(value.title)
        binding.itemLinkInput.setTextWithNoAnimation(value.link)
        binding.itemLoginInput.setTextWithNoAnimation(value.login)
        binding.itemEmailInput.setTextWithNoAnimation(value.email)
        binding.itemPhoneInput.setTextWithNoAnimation(value.phone)
        binding.itemPasswordInput.setTextWithNoAnimation(value.password)
        binding.itemPinInput.setTextWithNoAnimation(value.pin)
        binding.itemDetailsInput.setTextWithNoAnimation(value.details)
    }

    override fun initSaveButtonStateObserver() {
        initTextFillObserver(binding.itemTitleInput, CredentialsEditViewModel.INDEX_TITLE)
        initTextFillObserver(binding.itemLinkInput, CredentialsEditViewModel.INDEX_LINK)
        initTextFillObserver(binding.itemLoginInput, CredentialsEditViewModel.INDEX_LOGIN)
        initTextFillObserver(binding.itemEmailInput, CredentialsEditViewModel.INDEX_EMAIL)
        initTextFillObserver(binding.itemPhoneInput, CredentialsEditViewModel.INDEX_PHONE)
        initTextFillObserver(binding.itemPasswordInput, CredentialsEditViewModel.INDEX_PASSWORD)
        initTextFillObserver(binding.itemPinInput, CredentialsEditViewModel.INDEX_PIN)
    }
}
